{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example converting 10X data to Adaptive format\n",
    "\n",
    "This notebook demonstrates how to convert 10X T-cell receptor gene names to the Adaptive format."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import libraries and define filepaths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tcrconvert import convert\n",
    "\n",
    "dat10x_f = '/Users/emmabishop/workspace/tcrconvert/tcrconvert/data/example_10x.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load example data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACCACGA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGACCACGA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>521</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV1-2</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ12</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAVMDSSYKLIF</td>\n",
       "      <td>TGTGCTGTGATGGATAGCAGCTATAAATTGATCTTC</td>\n",
       "      <td>1569</td>\n",
       "      <td>2</td>\n",
       "      <td>clonotype16</td>\n",
       "      <td>clonotype16_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGACCACGA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGACCACGA-1_contig_3</td>\n",
       "      <td>True</td>\n",
       "      <td>584</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV6-1</td>\n",
       "      <td>TRBD2</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CASSGLAGGYNEQFF</td>\n",
       "      <td>TGTGCCAGCAGTGGACTAGCGGGGGGCTACAATGAGCAGTTCTTC</td>\n",
       "      <td>5238</td>\n",
       "      <td>7</td>\n",
       "      <td>clonotype16</td>\n",
       "      <td>clonotype16_consensus_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGCTCTTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGGCTCTTA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>551</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV6-4</td>\n",
       "      <td>TRBD2</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CASSGVAGGTDTQYF</td>\n",
       "      <td>TGTGCCAGCAGTGGGGTAGCGGGAGGCACAGATACGCAGTATTTT</td>\n",
       "      <td>3846</td>\n",
       "      <td>4</td>\n",
       "      <td>clonotype26</td>\n",
       "      <td>clonotype26_consensus_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGCTCTTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGGCTCTTA-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>518</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV1-2</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ33</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAVKDSNYQLIW</td>\n",
       "      <td>TGTGCTGTGAAGGATAGCAACTATCAGTTAATCTGG</td>\n",
       "      <td>2019</td>\n",
       "      <td>2</td>\n",
       "      <td>clonotype26</td>\n",
       "      <td>clonotype26_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGAGTGAACGC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGTGAACGC-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>674</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV2</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CASNQGLNYGYTF</td>\n",
       "      <td>TGTGCCAGCAATCAGGGCCTTAACTATGGCTACACCTTC</td>\n",
       "      <td>3002</td>\n",
       "      <td>6</td>\n",
       "      <td>clonotype81</td>\n",
       "      <td>clonotype81_consensus_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode  is_cell                    contig_id  high_confidence  \\\n",
       "0  AAACCTGAGACCACGA-1     True  AAACCTGAGACCACGA-1_contig_1             True   \n",
       "1  AAACCTGAGACCACGA-1     True  AAACCTGAGACCACGA-1_contig_3             True   \n",
       "2  AAACCTGAGGCTCTTA-1     True  AAACCTGAGGCTCTTA-1_contig_1             True   \n",
       "3  AAACCTGAGGCTCTTA-1     True  AAACCTGAGGCTCTTA-1_contig_2             True   \n",
       "4  AAACCTGAGTGAACGC-1     True  AAACCTGAGTGAACGC-1_contig_1             True   \n",
       "\n",
       "   length chain   v_gene d_gene   j_gene c_gene  full_length  productive  \\\n",
       "0     521   TRA  TRAV1-2   None   TRAJ12   TRAC         True        True   \n",
       "1     584   TRB  TRBV6-1  TRBD2  TRBJ2-1  TRBC2         True        True   \n",
       "2     551   TRB  TRBV6-4  TRBD2  TRBJ2-3  TRBC2         True        True   \n",
       "3     518   TRA  TRAV1-2   None   TRAJ33   TRAC         True        True   \n",
       "4     674   TRB    TRBV2  TRBD1  TRBJ1-2  TRBC1         True        True   \n",
       "\n",
       "              cdr3                                        cdr3_nt  reads  \\\n",
       "0     CAVMDSSYKLIF           TGTGCTGTGATGGATAGCAGCTATAAATTGATCTTC   1569   \n",
       "1  CASSGLAGGYNEQFF  TGTGCCAGCAGTGGACTAGCGGGGGGCTACAATGAGCAGTTCTTC   5238   \n",
       "2  CASSGVAGGTDTQYF  TGTGCCAGCAGTGGGGTAGCGGGAGGCACAGATACGCAGTATTTT   3846   \n",
       "3     CAVKDSNYQLIW           TGTGCTGTGAAGGATAGCAACTATCAGTTAATCTGG   2019   \n",
       "4    CASNQGLNYGYTF        TGTGCCAGCAATCAGGGCCTTAACTATGGCTACACCTTC   3002   \n",
       "\n",
       "   umis raw_clonotype_id         raw_consensus_id  \n",
       "0     2      clonotype16  clonotype16_consensus_1  \n",
       "1     7      clonotype16  clonotype16_consensus_2  \n",
       "2     4      clonotype26  clonotype26_consensus_2  \n",
       "3     2      clonotype26  clonotype26_consensus_1  \n",
       "4     6      clonotype81  clonotype81_consensus_2  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat10x = pd.read_csv(dat10x_f)\n",
    "dat10x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take a selection of columns and add a row of nonsense genes, to demonstrate what happens when you data isn't in the IMGT reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>cdr3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACCACGA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TRAV1-2</td>\n",
       "      <td>TRAJ12</td>\n",
       "      <td>CAVMDSSYKLIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGACCACGA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TRBV6-1</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>CASSGLAGGYNEQFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGCTCTTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TRBV6-4</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>CASSGVAGGTDTQYF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGCTCTTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TRAV1-2</td>\n",
       "      <td>TRAJ33</td>\n",
       "      <td>CAVKDSNYQLIW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGAGTGAACGC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TRBV2</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>CASNQGLNYGYTF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BARCODE</td>\n",
       "      <td>True</td>\n",
       "      <td>TRBVXYZ</td>\n",
       "      <td>TRBJ555</td>\n",
       "      <td>CDR3HERE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode is_cell   v_gene   j_gene             cdr3\n",
       "0  AAACCTGAGACCACGA-1    True  TRAV1-2   TRAJ12     CAVMDSSYKLIF\n",
       "1  AAACCTGAGACCACGA-1    True  TRBV6-1  TRBJ2-1  CASSGLAGGYNEQFF\n",
       "2  AAACCTGAGGCTCTTA-1    True  TRBV6-4  TRBJ2-3  CASSGVAGGTDTQYF\n",
       "3  AAACCTGAGGCTCTTA-1    True  TRAV1-2   TRAJ33     CAVKDSNYQLIW\n",
       "4  AAACCTGAGTGAACGC-1    True    TRBV2  TRBJ1-2    CASNQGLNYGYTF\n",
       "5             BARCODE    True  TRBVXYZ  TRBJ555         CDR3HERE"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df = dat10x[['barcode', 'is_cell', 'v_gene', 'j_gene', 'cdr3']].copy()\n",
    "my_df.loc[5] = ['BARCODE', 'True', 'TRBVXYZ', 'TRBJ555', 'CDR3HERE']\n",
    "my_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do the conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CONVERTING FROM 10X: CHOOSING *01 AS ALLELE FOR ALL GENES\n",
      "These genes are not in the IMGT reference and have replaced with NA. Please repair gene names manually and re-run:\n",
      " ['TRBVXYZ', 'TRBJ555']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>v_resolved</th>\n",
       "      <th>j_resolved</th>\n",
       "      <th>cdr3_amino_acid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACCACGA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TCRAV01-02*01</td>\n",
       "      <td>TCRAJ12*01</td>\n",
       "      <td>CAVMDSSYKLIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGACCACGA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TCRBV06-01*01</td>\n",
       "      <td>TCRBJ02-01*01</td>\n",
       "      <td>CASSGLAGGYNEQFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGCTCTTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TCRBV06-04*01</td>\n",
       "      <td>TCRBJ02-03*01</td>\n",
       "      <td>CASSGVAGGTDTQYF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGCTCTTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TCRAV01-02*01</td>\n",
       "      <td>TCRAJ33*01</td>\n",
       "      <td>CAVKDSNYQLIW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGAGTGAACGC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>TCRBV02*01</td>\n",
       "      <td>TCRBJ01-02*01</td>\n",
       "      <td>CASNQGLNYGYTF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BARCODE</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDR3HERE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode is_cell     v_resolved     j_resolved  cdr3_amino_acid\n",
       "0  AAACCTGAGACCACGA-1    True  TCRAV01-02*01     TCRAJ12*01     CAVMDSSYKLIF\n",
       "1  AAACCTGAGACCACGA-1    True  TCRBV06-01*01  TCRBJ02-01*01  CASSGLAGGYNEQFF\n",
       "2  AAACCTGAGGCTCTTA-1    True  TCRBV06-04*01  TCRBJ02-03*01  CASSGVAGGTDTQYF\n",
       "3  AAACCTGAGGCTCTTA-1    True  TCRAV01-02*01     TCRAJ33*01     CAVKDSNYQLIW\n",
       "4  AAACCTGAGTGAACGC-1    True     TCRBV02*01  TCRBJ01-02*01    CASNQGLNYGYTF\n",
       "5             BARCODE    True            NaN            NaN         CDR3HERE"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = convert.convert_tcr(my_df, fmt_from='tenx', fmt_to='adaptive', convert_cols=True)\n",
    "out"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
